From 0b6747887b4bc5c548c4301f35b1f8352098fd8e Mon Sep 17 00:00:00 2001
From: Huynh Thanh Hung <hung.huynh.wz@renesas.com>
Date: Wed, 4 Feb 2026 10:52:55 +0700
Subject: [PATCH 21/22] r9a09g047: Add FDP special power on/off configuration

FDP in R9A09G047 have special procedure power on/power off, then we need to
separate it with normal device

A. Procedure for starting FDP:
   1. Power on all the power.
   2. Supply all clocks:
   3. Confirm all clocks to started.
   4. Transition all Q-ch for all Q_RUN:
      -> Write 3333_3333h to the CPG_LP_FDP1_CTL register.
   5. Negate the reset_brdg_n.
   6. Confirm the reset_brdg_n was already negated.
   7. Negate the reset_fcpf_p and reset_fdp1_p.
   8. Confirm the reset_fcpf_p and reset_fdp1_p already negated.

B. Procedure for stopping FDP:
   1. Transition the Q-ch for all power supply control to Q_STOPPED:
      -> Write 0110_1111h to the CPG_LP_FDP1_CTL register.
   2. Confirm the Q-ch for all power supply control has transitioned to
Q_STOPPED:
      -> Read 0330_3333h from the CPG_LP_FDP1_CTL register.
   3. Assert reset_fcpf_p and reset_fdp1_p.
   4. Confirm reset_fcpf_p and reset_fdp1_p asserted.
   5. Assert reset_brdg_n.
   6. Confirm reset_brdg_n already asserted.
   7. Power off all the power.

Signed-off-by: Huynh Thanh Hung <hung.huynh.wz@renesas.com>
---
 vspm-module/files/vspm/drv/Makefile         |   1 +
 vspm-module/files/vspm/drv/r9a09g047_info.c | 295 ++++++++++++++++++++
 vspm-module/files/vspm/drv/vspm_main.c      |  13 +
 vspm-module/files/vspm/drv/vspm_main.h      |   6 +
 4 files changed, 315 insertions(+)
 create mode 100644 vspm-module/files/vspm/drv/r9a09g047_info.c

diff --git a/vspm-module/files/vspm/drv/Makefile b/vspm-module/files/vspm/drv/Makefile
index a6a06bb..ae403b5 100644
--- a/vspm-module/files/vspm/drv/Makefile
+++ b/vspm-module/files/vspm/drv/Makefile
@@ -4,6 +4,7 @@ CFILES += manager/vspm_control.c manager/vspm_drv_vsp.c manager/vspm_drv_fdp.c m
 CFILES += manager/vspm_job_manager.c manager/vspm_lib.c manager/vspm_sort_queue.c manager/vspm_task.c
 CFILES += vsp/vsp_drv.c vsp/vsp_drv_phy.c vsp/vsp_drv_par.c
 CFILES += fdp/fdp_drv.c fdp/fdp_drv_l.c
+CFILES += r9a09g047_info.c
 
 obj-m += vspm.o
 vspm-objs := $(CFILES:.c=.o)
diff --git a/vspm-module/files/vspm/drv/r9a09g047_info.c b/vspm-module/files/vspm/drv/r9a09g047_info.c
new file mode 100644
index 0000000..17a189a
--- /dev/null
+++ b/vspm-module/files/vspm/drv/r9a09g047_info.c
@@ -0,0 +1,295 @@
+/*************************************************************************/ /*
+ * R9A09G047 Device Specification
+ *
+ * Copyright (C) 2026 Renesas Electronics Corporation
+ *
+ * License        Dual MIT/GPLv2
+ *
+ * The contents of this file are subject to the MIT license as set out below.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * Alternatively, the contents of this file may be used under the terms of
+ * the GNU General Public License Version 2 ("GPL") in which case the provisions
+ * of GPL are applicable instead of those above.
+ *
+ * If you wish to allow use of your version of this file only under the terms of
+ * GPL, and not to allow others to use your version of this file under the terms
+ * of the MIT license, indicate your decision by deleting the provisions above
+ * and replace them with the notice and other provisions required by GPL as set
+ * out in the file called "GPL-COPYING" included in this distribution. If you do
+ * not delete the provisions above, a recipient may use your version of this
+ * file under the terms of either the MIT license or GPL.
+ *
+ * This License is also included in this distribution in the file called
+ * "MIT-COPYING".
+ *
+ * EXCEPT AS OTHERWISE STATED IN A NEGOTIATED AGREEMENT: (A) THE SOFTWARE IS
+ * PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING
+ * BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+ * PARTICULAR PURPOSE AND NONINFRINGEMENT; AND (B) IN NO EVENT SHALL THE AUTHORS
+ * OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+ * WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR
+ * IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ *
+ *
+ * GPLv2:
+ * If you wish to use this file under the terms of GPL, following terms are
+ * effective.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; version 2 of the License.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ */ /*************************************************************************/
+
+#include <linux/platform_device.h>
+#include <linux/pm_runtime.h>
+#include "vspm_ip_ctrl.h"
+#include "fdp_drv.h"
+#include "vspm_main.h"
+#include "vspm_log.h"
+
+#define STATUS_LOOP_TIME			(2)
+#define STATUS_LOOP_CNT				(500)
+
+/* define low-power-consumption sequence control register offset */
+#define CPG_LP_FDP1_CTL				(0x0CA4)
+#define CPG_LP_FDP1_PWRREQ			(BIT(16) | BIT(28))
+#define CPG_LP_FDP1_PWRACC			(BIT(17) | BIT(29))
+#define CPG_LP_FDP1_CLKREQ_S		(BIT(0) | BIT(8) | BIT(20))
+#define CPG_LP_FDP1_CLKREQ_M		(BIT(4) | BIT(12) | BIT(24))
+#define CPG_LP_FDP1_CLKREQ			(CPG_LP_FDP1_CLKREQ_S | CPG_LP_FDP1_CLKREQ_M)
+#define CPG_LP_FDP1_CLKACC_S		(BIT(1) | BIT(9) | BIT(21))
+#define CPG_LP_FDP1_CLKACC_M		(BIT(5) | BIT(13) | BIT(25))
+#define CPG_LP_FDP1_CLKACC			(CPG_LP_FDP1_CLKACC_S | CPG_LP_FDP1_CLKACC_M)
+
+int syscon_set_sync(struct regmap *syscon, unsigned int offset, unsigned int value, unsigned int check, int mask)
+{
+	unsigned int	reg_data	= 0;
+	unsigned int	loopcnt		= 0;
+
+	if (!syscon)
+		return -EPIPE;
+
+	if (regmap_write(syscon, offset, value)) {
+		EPRINT("%s: failed to write to syscon\n", __func__);
+		return -EAGAIN;
+	}
+
+	if (!mask)
+		return 0;
+
+	/* confirm Q-ch for all power supply control has transitioned */
+	loopcnt = STATUS_LOOP_CNT;
+	do {
+		if (regmap_read(syscon, offset, &reg_data)) {
+			EPRINT("%s: read error \n",__func__);
+			return -EAGAIN;
+		}
+		msleep(STATUS_LOOP_TIME);
+	} while (((reg_data & mask) != (check & mask)) &&
+			(--loopcnt > 0));
+
+	if (loopcnt == 0) {
+		EPRINT("%s: timeout waiting for power supply control transition to 0x%x (cur:0x%x, set:0x%x, mask:0x%x)" \
+				"\n", __func__, check, reg_data, value, mask);
+		return -EAGAIN;
+	}
+	return 0;
+}
+
+static int clock_disable(struct vspm_device *vdev)
+{
+	int	i	= 0;
+
+	if (!vdev)
+		return -EINVAL;
+
+	for (i = (vdev->num_clks - 1); i >= 0 ; i--) {
+		if (vdev->clks[i]) {
+			clk_disable_unprepare(vdev->clks[i]);
+		}
+	}
+	return 0;
+}
+
+static int clock_enable(struct vspm_device *vdev)
+{
+	int	i	= 0;
+	int	ret	= -EINVAL;
+
+	if (!vdev)
+		return ret;
+
+	for (i = 0 ; i < vdev->num_clks ; i++) {
+		if (vdev->clks[i]) {
+			ret = clk_prepare_enable(vdev->clks[i]);
+			if (ret) {
+				EPRINT("%s: Fail to enable clock %d \n", __func__, i);
+				return ret;
+			}
+		}
+	}
+	return ret;
+}
+
+static int reset_assert(struct vspm_device *vdev)
+{
+	int	i		= 0;
+	int	loopcnt	= 0;
+
+	if (!vdev)
+		return -EINVAL;
+
+	for (i = (vdev->num_rsts - 1); i >= 0 ; i--) {
+		if (reset_control_assert(vdev->rsts[i])) {
+			EPRINT("%s: failed to assert resets!! \n", __func__);
+			return -EAGAIN;
+		}
+		/* waiting assert process */
+		loopcnt = STATUS_LOOP_CNT;
+		do {
+			/* sleep */
+			msleep(STATUS_LOOP_TIME);
+		} while ((!reset_control_status(vdev->rsts[i])) &&
+				(--loopcnt > 0));
+
+		if (loopcnt == 0) {
+			EPRINT("%s: happen to timeout after assert resets!! \n", __func__);
+			return -EAGAIN;
+		}
+	}
+	return 0;
+}
+
+static int reset_deassert(struct vspm_device *vdev)
+{
+	int	i		= 0;
+	int	loopcnt	= 0;
+
+	if (!vdev)
+		return -EINVAL;
+
+	for (i = 0; i < vdev->num_rsts ; i++) {
+		if (reset_control_deassert(vdev->rsts[i])) {
+			EPRINT("%s: failed to deassert resets!! \n", __func__);
+			return -EAGAIN;
+		}
+		/* waiting deassert process */
+		loopcnt = STATUS_LOOP_CNT;
+		do {
+			/* sleep */
+			msleep(STATUS_LOOP_TIME);
+		} while ((reset_control_status(vdev->rsts[i])) &&
+				(--loopcnt > 0));
+
+		if (loopcnt == 0) {
+			EPRINT("%s: happen to timeout after deassert resets!! \n", __func__);
+			return -EAGAIN;
+		}
+	}
+	return 0;
+}
+
+static int r9a09g047_fdp_power_on(struct device *dev)
+{
+	struct platform_device	*pdev	= to_platform_device(dev);
+	struct vspm_device		*vdev	= platform_get_drvdata(pdev);
+	int						ercd 	= -EAGAIN;
+
+	if (!vdev || !vdev->syscon_cpg)
+		return -EINVAL;
+
+	/* A2 - A3. Supply all clocks */
+	ercd = clock_enable(vdev);
+	if (ercd) {
+		EPRINT("%s: Cannot enable clock: %d \n",__func__, ercd);
+		return ercd;
+	}
+
+	/* W/A: Deassert resets should be done before supplying clock */
+	/* A5 - A6 - A7 - A8. Deassert resets */
+	ercd = reset_deassert(vdev);
+	if (ercd) {
+		EPRINT("%s: Cannot deassert resets \n",__func__);
+		goto err1;
+	}
+
+	/* A4. Release lower power state */
+	ercd = syscon_set_sync(vdev->syscon_cpg, CPG_LP_FDP1_CTL,
+							CPG_LP_FDP1_PWRREQ | CPG_LP_FDP1_CLKREQ,
+							CPG_LP_FDP1_PWRREQ | CPG_LP_FDP1_CLKREQ,
+							CPG_LP_FDP1_PWRREQ | CPG_LP_FDP1_CLKREQ /* W/A: Not recheck the acc bit */
+							);
+	if (ercd) {
+		EPRINT("%s: Cannot release lower power state: %d\n",__func__, ercd);
+		goto err2;
+	}
+
+	return ercd;
+err2:
+	reset_assert(vdev);
+err1:
+	clock_disable(vdev);
+	return ercd;
+}
+
+static int r9a09g047_fdp_power_off(struct device *dev)
+{
+	struct platform_device	*pdev	= to_platform_device(dev);
+	struct vspm_device		*vdev	= platform_get_drvdata(pdev);
+	int						ercd	= -EAGAIN;
+
+	if (!vdev || !vdev->syscon_cpg)
+		return -EINVAL;
+
+	/* B1 - B2. Make lower power state */
+	ercd = syscon_set_sync(vdev->syscon_cpg, CPG_LP_FDP1_CTL,
+							(unsigned int)(~(CPG_LP_FDP1_PWRREQ | CPG_LP_FDP1_PWRACC)) & (CPG_LP_FDP1_CLKREQ | CPG_LP_FDP1_CLKACC),
+							0, /* Only check the Power Request/Acc bit */
+							CPG_LP_FDP1_PWRREQ | CPG_LP_FDP1_PWRACC	/* Check supply request/acc bit as document said */
+							);
+	if (ercd) {
+		EPRINT("%s: Cannot make lower power state: %d \n", __func__, ercd);
+		return ercd;
+	}
+	/* B3 - B4 - B5 - B6. Assert resets */
+	ercd = reset_assert(vdev);
+	if (ercd) {
+		EPRINT("%s: Cannot assert reset: %d \n",__func__, ercd);
+		goto err1;
+	}
+	/* Disable clock */
+	ercd = clock_disable(vdev);
+	if (ercd) {
+		EPRINT("%s: Cannot disable clock: %d \n",__func__, ercd);
+		goto err2;
+	}
+	return 0;
+err2:
+	reset_deassert(vdev);
+err1:
+	syscon_set_sync(vdev->syscon_cpg, CPG_LP_FDP1_CTL,
+					CPG_LP_FDP1_PWRREQ | CPG_LP_FDP1_CLKREQ,
+					CPG_LP_FDP1_PWRREQ, 0);
+	return ercd;
+}
+
+const struct vspm_device_ops r9a09g047_fdp_info = {
+	.power_on			= r9a09g047_fdp_power_on,
+	.power_off			= r9a09g047_fdp_power_off,
+	.require_features	= VSPM_DT_SYSCON_CPG,
+};
diff --git a/vspm-module/files/vspm/drv/vspm_main.c b/vspm-module/files/vspm/drv/vspm_main.c
index b4e335d..157ee91 100644
--- a/vspm-module/files/vspm/drv/vspm_main.c
+++ b/vspm-module/files/vspm/drv/vspm_main.c
@@ -64,6 +64,8 @@
 #include <linux/io.h>
 #include <linux/pm_runtime.h>
 #include <linux/reset.h>
+#include <linux/regmap.h>
+#include <linux/mfd/syscon.h>
 
 #include "vspm_public.h"
 #include "vspm_ip_ctrl.h"
@@ -78,6 +80,7 @@ static const struct soc_device_attribute rz_vsp_quirk_attrs[] = {
 	{ /* sentinel */ }
 };
 static const struct soc_device_attribute rz_fdp_quirk_attrs[] = {
+	{ .soc_id = "r9a09g047", .data = &r9a09g047_fdp_info },
 	{ /* sentinel */ }
 };
 
@@ -412,6 +415,7 @@ static int vspm_fdp_probe(struct platform_device *pdev)
 	int ch;
 	struct vspm_device *vdev = NULL;
 	const struct soc_device_attribute *attr;
+	struct regmap *syscon = NULL;
 
 	/* get channel */
 	of_property_read_u32(np, "renesas,#ch", &ch);
@@ -467,6 +471,15 @@ static int vspm_fdp_probe(struct platform_device *pdev)
 	}
 	vdev->num_rsts = RSTNUM;
 
+	if (vdev->ops && vdev->ops->require_features & VSPM_DT_SYSCON_CPG) {
+		syscon = syscon_regmap_lookup_by_phandle(np, "renesas,syscon-cpg");
+		if (IS_ERR(syscon)) {
+			EPRINT("failed to get syscon regmap for CPG\n");
+			return -1;
+		}
+		vdev->syscon_cpg = syscon;
+	}
+
 	/* set driver data */
 	platform_set_drvdata(pdev, vdev);
 	pdrv->fdp_pdev[ch] = pdev;
diff --git a/vspm-module/files/vspm/drv/vspm_main.h b/vspm-module/files/vspm/drv/vspm_main.h
index 25ea2da..0c6ce59 100644
--- a/vspm-module/files/vspm/drv/vspm_main.h
+++ b/vspm-module/files/vspm/drv/vspm_main.h
@@ -62,6 +62,7 @@
 #include <linux/fs.h>
 #include <linux/clk.h>
 #include <linux/reset.h>
+#include <linux/regmap.h>
 
 #define DEVNAME				"vspm"
 #define DEVNUM				1
@@ -83,9 +84,12 @@
 #define CLKNUM				3
 #define RSTNUM				3
 
+#define VSPM_DT_SYSCON_CPG	(BIT(0))
+
 struct vspm_device_ops {
 	int (*power_on)(struct device *dev);
 	int (*power_off)(struct device *dev);
+	int require_features;
 };
 
 struct vspm_device {
@@ -95,6 +99,7 @@ struct vspm_device {
 	char rsts_names[255][RSTNUM];
 	int num_clks;
 	int num_rsts;
+	struct regmap* syscon_cpg;
 	const struct vspm_device_ops *ops;
 };
 
@@ -141,4 +146,5 @@ long vspm_suspend(struct vspm_drvdata *pdrv);
 long vspm_resume(struct vspm_drvdata *pdrv);
 long vspm_cancel(struct vspm_privdata *priv);
 
+extern const struct vspm_device_ops r9a09g047_fdp_info;
 #endif /* __VSPM_MAIN_H__ */
-- 
2.25.1

